### PR TITLE
fix(api-service,dashboard): Crate of fixes for variable suggestions

### PR DESCRIPTION
**Pull request by SokratisVidros**
### What changed? Why was the change needed?

https://linear.app/novu/issue/NV-5104/adding-variable-on-the-fly-in-a-button-is-not-working
https://linear.app/novu/issue/NV-5105/when-trying-to-add-variables-on-fly-from-button-or-image-block-link

### CODE REVIEW

**Review by scopsy — APPROVED** [review]

File: apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts [hunk 1]
```diff
@@ -16,26 +15,13 @@ export class BuildPayloadSchema {
 
   @InstrumentUsecase()
   async execute(command: BuildPayloadSchemaCommand): Promise<JSONSchemaDto> {
-    const controlValues = await this.buildControlValues(command);
+    const controlValues = await this.getControlValues(command);
+    const sanitizedControlValues = await this.sanitizeControlValues(controlValues);
```
Comment (djabarovgeorge) [inline apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts]: i am not sure if the naming is correct here we do not sanitizeControlValues we create templateVariableNames and the function is returning not control value but string[] of variables

File: apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts [hunk 2]
```diff
@@ -87,4 +73,30 @@ export class BuildPayloadSchema {
 
     return processedValue;
   }
+
+  private async buildVariablesSchema(variables: string[]) {
```
Comment (djabarovgeorge) [inline apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts:77]: i think we lost the case of

 if (!controlValues.length) {
    return this.buildVariablesSchema(sanitizedControlValues);
      return emptyJsonSchema();	
    }
    
    
   if i remember correctly this case was meant to cover the mini preview on step create, before we create the initial control values

File: apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts [hunk 3]
```diff
@@ -87,4 +73,30 @@ export class BuildPayloadSchema {
...
+
+    const schema: JSONSchemaDto = {
+      type: 'object',
+      properties: {},
+      required: [],
+      additionalProperties: true,
+    };
+
+    if (variablesObject) {
+      for (const [key, value] of Object.entries(variablesObject)) {
```
Comment (djabarovgeorge) [inline apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts:92]: will this for loop cover nested payload as well, for example, `payload.bar.data.foo...` etc..?

File: apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts (hunk 2 above)
Comment (SokratisVidros) [reply apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts:77]: If I am not mistaken, this should be covered as it as the functions can all handle empty variables.

File: apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts (hunk 3 above)
Comment (SokratisVidros) [reply apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts:92]: Note, this is the reason e2e is failing, and I am pushing a fix.

File: apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts (hunk 1 above)
Comment (SokratisVidros) [reply apps/api/src/app/workflows-v2/usecases/build-payload-schema/build-payload-schema.usecase.ts]: Fixed

**Pull request merged**